### PR TITLE
Check for BUILD_TESTING and BUILD_SAMPLES inside Storage's cmakefiles

### PR DESCRIPTION
**Pull request by antkmsft**
New description:
In order for us to be able to compile a seperate lib without loading the entire cmake tree, individual lib cmakefiles needs to be buildable. It is what we need to do when bulding for vcpkg. But when we build just the lib, we need to ignore tests and samples.

Previously I tried to move tests and samples out of the lib cmakefile, but that method did not find understanding among the Storage team right away, so, after all, moving to the separate cmakefile is not the only way to achieve what we need, so recent commit is another take on that.

Old description:
It is good to keep one target/directory per cmakefile. Not only it is easier to read and maintain, but it is easier to build in isolation, which becomes very apparent as we look towards vcpkg integration and want to only produce a single library during the build.

--
This change was extracted from #1013, in order to isolate some of the changes.

### CODE REVIEW

**Comment by Jinming-Hu** [comment]
> Not only it is easier to read and maintain, but it is easier to build in isolation

I doubt that, I think one target is easier to maintain than 4 (more in the future).

> which becomes very apparent as we look towards vcpkg integration and want to only produce a single library during the build.

We don't build tests or samples when installing via vcpkg. Does current CMakeFiles cause any problems?

**Review by Jinming-Hu — COMMENTED** [review]
This is a very big change to storage code, please hold on merging until get storage team's approval.

File: sdk/storage/CMakeLists.txt
Comment (RickWinter) [inline sdk/storage/CMakeLists.txt:41]:
```suggestion
endif()

```

File: sdk/storage/azure-storage-files-shares/test/CMakeLists.txt
Comment (RickWinter) [inline sdk/storage/azure-storage-files-shares/test/CMakeLists.txt:4]: Can this be dialed back to version 3.5?

Comment (vhvb1989) [reply sdk/storage/azure-storage-files-shares/test/CMakeLists.txt:4]: Not yet.
We would also need to break the Storage test generation into multiple exe files instead of just one in other to go back to previous CMake versions.
Currently, the main storage test exe is created in the main CMake file with:
`add_executable(azure-storage-test)` without adding any source files.
Then, later, each subdirectory adds source files to this exe target.  This is only supported on CMake version 3.13.

We can either list all the source files in the main CMake file (uggly solution), or let each subdirectory to create its own exe file.

That could be a different PR for the future as an effort to move CMake supported version back

**Comment by antkmsft** [comment]
> We don't build tests or samples when installing via vcpkg. Does current CMakeFiles cause any problems?
We currently don't build as vcpkg. I have a PR to enable it, this PR was extracted from it:  #1013 
When we build vcpkg, we have to build a single directory, so library needs to have a single cmakefile of its own.
More than that, storage's tests is actualy a single target spread across 4 corresponding cmakefiles. These things should exist in parallel dimension of cmakefiles, just as core does.

**Comment by antkmsft** [comment]
@katmsft , @vinjiang asking for your review per Jinming's request.

**Comment by Jinming-Hu** [comment]
> > We don't build tests or samples when installing via vcpkg. Does current CMakeFiles cause any problems?
> 
> We currently don't build as vcpkg. I have a PR to enable it, this PR was extracted from it: #1013
> When we build vcpkg, we have to build a single directory, so library needs to have a single cmakefile of its own.
> More than that, storage's tests is actualy a single target spread across 4 corresponding cmakefiles. These things should exist in parallel dimension of cmakefiles, just as core does.

My initial idea was that our customers can build any of these four libraries as long as they there's no dependency on each other. So I added four CMakeFiles in four subdirectories. But for tests, I just didn't see the necessity of separating them into four executables (and still not after seeing your PR). So can you explain how can we as developers or our customers benefit from this PR? Because I did see extra maintenance cost from developer's perspective, but I'd like to make some sacrifice as long as it brings convenience to our customers.

**Comment by antkmsft** [comment]
> My initial idea was that our customers can build any of these four libraries as long as they there's no dependency on each other. So I added four CMakeFiles in four subdirectories. But for tests, I just didn't see the necessity of separating them into four executables (and still not after seeing your PR). So can you explain how can we as developers or our customers benefit from this PR? Because I did see extra maintenance cost from developer's perspective, but I'd like to make some sacrifice as long as it brings convenience to our customers.

And now imagine, we are installing a single package `azure-storage-blobs`. For it to be buildable independent package, we need to build by loading a cmakefile in the library folder. For the package, we only build the library, and not tests. Tests have extra dependencies. How do we distinguish that type of build from the build when we build from VisualStudio as we do today.

BTW, if you load the cmakefile at the top folder, everything works exactly the same as today, all the same stuff is built. It is only when you load the cmakefile that lives in the `azure-storage-blobs`, we would build the `azure-storage-blobs`. Note that today it is not even possible to load the cmakefile from the `azure-storage-blobs` directory and build successfully. I have a PR that enables that - #1013. But this PR is a step in that direction, and having straightforward makefiles with clear goals and targets is a prerequisite for turning our libraries into VcPkgs.

And Core have this separation of lib, samples, and tests, and there is no problem. I would even argue that for the sample, it is better to have an isolated cmakefile, because it shows the user how to consume the library via cmake. You can copy the sample directory out of the SDK build tree, and build, and it is self-sufficient.

We also used this approach in our other SDK for C, there were no problems with maintenance.

File: sdk/storage/azure-storage-blobs/test/CMakeLists.txt
Comment (vhvb1989) [inline sdk/storage/azure-storage-blobs/test/CMakeLists.txt:7]: This won't work from here.
Please try to generate code coverage reports locally to make sure it would still work after moving this here.

Having this here means that the test exe will be compiled with the flags for profiling, like if we were wanting the coverage for the test code.

**Review by vhvb1989 — APPROVED** [review]
Make sure code coverage still works after this before merging.

Also please wait for Storage's team approval before merging. Maybe set/ask some ETA on waiting for them.

I am really happy for this change, btw, thank you for taking care of it!

**Comment by antkmsft** [comment]
@Jinming-Hu , I moved tests and samples back to the lib source cmakefiles, would you approve this approach?

**Comment by Jinming-Hu** [comment]
> @Jinming-Hu , I moved tests and samples back to the lib source cmakefiles, would you approve this approach?

Yes, absolutely. Actually storage team talked about this PR today in the internal sync meeting today. We understand that this PR is blocking your work here, and I'm going to take a look, hopefully on Tuesday or Wednesday. Thank you for your work.

**Review by Jinming-Hu — APPROVED** [review]
I don't know if this will work. As @vhvb1989 said, should we make sure code coverage still works before merging? Or we can get back to it later.

**Comment by antkmsft** [comment]
> I don't know if this will work. As @vhvb1989 said, should we make sure code coverage still works before merging? Or we can get back to it later.

Victor's point was about when things were done another way (but it still was possible to resolve by pulling the code coverage command into the lib cmakefile). In this case, we only introducing additional internal check on having the sample and test files included, instead of relying on out-of-file check only, and it should make no difference to code coverage.

**Pull request merged**
